File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    queue = deque([(0,0,0)])    
    visited = set([(0,0)])
    dxs, dys = (-1,0,1,0),(0,1,0,-1)
    n = len(maps)
    m = len(maps[0])
    
    def in_range(x,y):
        return 0<=x<n and 0<=y<m and (x,y) not in visited and maps[x][y] == 1
    
    while queue:
        x,y,cnt = queue.popleft()
        for dx,dy in zip(dxs,dys):
            nx , ny = x + dx, y + dy
            if in_range(nx,ny):
                if (nx,ny) == (n-1,m-1) :
                    return cnt + 2
                queue.append((nx,ny,cnt+1))
                visited.add((nx,ny))
                
    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[] dx = {-1,0,1,0};
    int[] dy = {0,1,0,-1};
    
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        
        Queue<Robot> queue = new ArrayDeque<>();
        queue.offer(new Robot(0,0,1));
        
        boolean[][] visited = new boolean[n][m];
        visited[0][0] = true;
        
        while (!queue.isEmpty()){
            Robot r = queue.poll();
            if (r.x == n-1 && r.y == m-1){
                return r.moved;
            }
            
            for (int i =0; i< 4; i++){
                int nx = r.x + dx[i];
                int ny = r.y + dy[i];
                boolean inRange = 0<= nx && 0<= ny && n > nx && m > ny;
                if (inRange && !visited[nx][ny] && maps[nx][ny] == 1){
                    queue.offer(new Robot(nx,ny, r.moved + 1 ));
                    visited[nx][ny] = true;
                }
            }
                
        }
        
        return -1;
    }
    
    class Robot{
        int x;
        int y;
        int moved;
        Robot(int x, int y,int moved){
            this.x = x;
            this.y = y;
            this.moved = moved;
        }
    }
}